### PR TITLE
feat: add `Type.JvmToEff`

### PR DESCRIPTION
**Pull request by magnus-madsen**

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/phase/typer/ConstraintGen.scala [hunk 1]
```diff
@@ -766,40 +766,42 @@ object ConstraintGen {
...
         // --------------------------------------------------------
-        // Γ ⊢ new k(e₁ ...) : k
+        // Γ ⊢ new k(e₁ ...) : k \ JvmToEff[ι]
         val clazzTpe = Type.getFlixType(clazz)
         val (tpes, effs) = exps.map(visitExp).unzip
-        val baseEffs = BaseEffects.of(clazz, loc)
-        c.unifyType(jvar, Type.UnresolvedJvmType(Type.JvmMember.JvmConstructor(clazz, tpes), loc), loc) // unify constructor
-        c.unifyType(evar, Type.mkUnion(Type.IO :: baseEffs :: effs, loc), loc) // unify effects
+        c.unifyType(jvar, Type.UnresolvedJvmType(Type.JvmMember.JvmConstructor(clazz, tpes), loc), loc)
+        c.unifyType(jevar, Type.JvmToEff(jvar, loc), loc)
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/typer/ConstraintGen.scala:776]: Can you try it without jevar? I'm 76% sure it's unneeded.

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/typer/ConstraintGen.scala:776]: It immediately crashes

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/typer/ConstraintGen.scala:776]: a.uwaterloo.flix.util.InternalCompilerException: Unexpected type: 'JvmToEff(iterator)'.
	at ca.uwaterloo.flix.language.phase.unification.BoolFormula$.fromEffType(BoolFormula.scala:144)
	at ca.uwaterloo.flix.language.phase.unification.BoolFormula$.fromEffType(BoolFormula.scala:138)
	at ca.uwaterloo.flix.language.phase.unification.TypeMinimization$.minimizeBoolAlg(TypeMinimization.scala:106)
	at ca.uwaterloo.flix.language.phase.unification.TypeMinimization$.minimizeType(TypeMinimization.scala:35)
	at ca.uwaterloo.flix.language.phase.typer.ConstraintSolver$.resolveOne(ConstraintSolver.scala:265)
	at ca.uwaterloo.flix.language.phase.typer.ConstraintSolver$.tryResolve$1(ConstraintSolver.scala:246)
	at ca.uwaterloo.flix.language.phase.typer.ConstraintSolver$.resolveOneOf(ConstraintSolver.scala:256)
	at ca.uwaterloo.flix.language.phase.typer.ConstraintSolver$.resolve(ConstraintSolver.scala:222)
	at ca.uwaterloo.flix.language.phase.typer.ConstraintSolver$.visitSpec(ConstraintSolver.scala:104)
	at ca.uwaterloo.flix.language.phase.typer.ConstraintSolver$.visitDef(ConstraintSolver.scala:50)

File: main/src/ca/uwaterloo/flix/language/phase/typer/BaseEffects.scala [hunk 2]
```diff
@@ -17,40 +17,72 @@ package ca.uwaterloo.flix.language.phase.typer
...
-    */
-  def of(m: Method, loc: SourceLocation): Type = {
-    val effs = methodEffs.getOrElse(m, Set.empty).toList
-    val tpes = effs.map(sym => Type.Cst(TypeConstructor.Effect(sym), loc))
-    Type.mkUnion(tpes, loc)
+   * Returns the base effects of calling the given method `m`.
+   */
+  def getMethodEffs(m: Method, loc: SourceLocation): Type = methodEffs.get(m) match {
+    case None =>
+      // Case 1: No effects for the constructor. Try the class map.
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/typer/BaseEffects.scala:72]: `s/constructor/method/`

File: main/src/ca/uwaterloo/flix/language/phase/unification/BoolFormula.scala [hunk 3]
```diff
@@ -197,40 +202,35 @@ object BoolFormula {
...
   }
 
   /**
-    * Union of variable and effect types.
+    * An irreducible effect.
     */
-  sealed trait VarOrEff
+  sealed trait IrreducibleEff
+
+  object IrreducibleEff {
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/unification/BoolFormula.scala:225]: IMO we should generally split renamings like this into a separate PR since they tend to distract from the real content.

**Review by mlutze — APPROVED** [review]

File: main/src/ca/uwaterloo/flix/language/ast/Type.scala [hunk 4]
```diff
@@ -110,9 +115,26 @@ sealed trait Type {
...
 
     case Type.JvmToType(tpe, _) => tpe.assocs
+    case Type.JvmToEff(tpe, _) => tpe.assocs
     case Type.UnresolvedJvmType(member, _) => member.getTypeArguments.foldLeft(Set.empty[Type.AssocType])((acc, t) => acc ++ t.assocs)
   }
 
+  /**
+   * Returns all the JvmToEffs in the given type.
+   */
+  def jvmToEffs: Set[Type.JvmToEff] = this match {
```
[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/Type.scala:125]: Its not great, I agree.

File: main/src/ca/uwaterloo/flix/language/phase/unification/BoolFormula.scala (hunk 3 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/unification/BoolFormula.scala:225]: True.
